spec/lib/catalog/submit_order_item_spec.rb

```diff
@@ -54,18 +54,36 @@
     end
 
     context "when the source is valid" do
-      it "updates the order item with the task id" do
-        expect(submit_order.process.order.order_items.first.topology_task_ref).to eq("100")
+      context "when pre order item has not restarted" do
+        before { pre_order_item.update(:state => 'Created') }
+
+        it 'orders the pre item and gets back the task id' do
+          expect(submit_order.process.order.order_items.first.topology_task_ref).to eq("100")
+        end
+      end
+
+      context "when pre order item has completed" do
+        it "orders the order item and gets back the task id" do
+          expect(submit_order.process.order.order_items.second.topology_task_ref).to eq("100")
+        end
+      end
+
+      context "when order item has completed" do
+        before { order_item.update(:state => 'Completed') }
+
+        it 'orders the post item and gets back the task id' do
+          expect(submit_order.process.order.order_items.last.topology_task_ref).to eq("100")
+        end
```

Comment:
The change to `SubmitOrderItem` was actually minimal, mainly just one line https://github.com/RedHatInsights/catalog-api/pull/805/files#diff-374ebf6b55f9aa4412c8a8f1e144f8c8L17. We remove the loop by just submitting one item.

Yes. I will stub the `can_order?` method

Reply:
Right, I understand the change was simple, but how the system interacts as a whole is a lot different. Each of the tasks that are going to be coming back are separate task ids, and each portfolio item will have a different service offering ref. It would be very strange (although I guess technically possible) to have the same product as the before, applicable, and after.

My point was that the spec doesn't really have any indication of it being this complicated, and so I feel like we may be missing some coverage. Breaking the coverage out into a bunch of separate cases and stubbing everything will quickly become large, and the larger the test case, the more I think it presents a good opportunity to simplify and refactor. I don't have any qualms with how the code actually works, I think it will do the job nicely, I'm just concerned about potential future technical debt.

Reply:
Somehow I couldn't stub `can_order?` on the element of `order_item` array. The stub is lost when instance changed.